src/services/Provider.ts

```diff
@@ -765,16 +786,30 @@ export class Provider {
     payload.signature = signature
     payload.nonce = nonce
     payload.environment = computeEnv
+    payload.maxJobDuration = maxJobDuration
     payload.resources = resources
-    payload.chainId = chainId
     // kept for backwards compatibility (tests running against existing provider)
     payload.dataset = datasets[0]
     // new field for C2D v2
     payload.datasets = datasets
     payload.algorithm = algorithm
+    let chainIdCompute: number
+    if (chainId) {
+      chainIdCompute = chainId
+    } else {
+      chainIdCompute = await consumer.getChainId()
+    }
```

Comment:
nullish coalescing can be used for a cleaner conditional assignment.
```suggestion
    const chainIdCompute = chainId ?? await consumer.getChainId()
```